delta_Go/game_mechanics.py

```diff
@@ -0,0 +1,260 @@
+import pickle
+import random
+import sys
+import time
+from copy import copy, deepcopy
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, Callable, Dict, Optional
+
+import numpy as np
+import torch
+from gym.spaces import Box, Discrete
+from pygame import Surface
+from torch import nn
+
+from pettingzoo.classic import go_v5
+from pettingzoo.utils import BaseWrapper
+
+HERE = Path(__file__).parent.resolve()
+
+# Hack as this won't pip install on replit
+sys.path.append(str(HERE / "PettingZoo"))
+
+
+BOARD_SIZE = 9
+ALL_POSSIBLE_MOVES = np.arange(BOARD_SIZE**2 + 1)
+
+
+# The komi to use is much debated. 7.5 seems to
+# generalise well for different board sizes
+# lifein19x19.com/viewtopic.php?f=15&t=17750
+# 7.5 is also the komi used in alpha-go vs Lee Sedol
+# (non-integer means there are no draws)
+
+KOMI = 7.5
+
+
+def play_go(
+    your_choose_move: Callable,
+    opponent_choose_move: Callable,
+    game_speed_multiplier=1,
+    render=True,
+    verbose=False,
+) -> None:
+
+    env = GoEnv(
+        opponent_choose_move,
+        verbose=verbose,
+        render=render,
+        game_speed_multiplier=game_speed_multiplier,
+    )
+
+    observation, reward, done, info = env.reset()
+    done = False
+    while not done:
+        action = your_choose_move(observation, info["legal_moves"], env=env)
+        observation, reward, done, info = env.step(action)
+    if render:
+        time.sleep(100)
+    return reward
+
+
+class DeltaEnv(BaseWrapper):
+    def __init__(
+        self,
+        env,
+        opponent_choose_move: Callable,
+        verbose: bool = False,
+        render: bool = False,
+        game_speed_multiplier: int = 1,
+    ):
+
+        super().__init__(env)
+
+        self.opponent_choose_move = opponent_choose_move
+        self.render = render
+        self.verbose = verbose
+        self.game_speed_multiplier = game_speed_multiplier
+        self.action_space = Discrete(BOARD_SIZE**2 + 1)
+        self.observation_space = Box(low=-1, high=1, shape=(BOARD_SIZE, BOARD_SIZE))
+        self.num_envs = 1
+
+        # Which color do we play as
+        self.player = random.choice(["black_0", "white_0"])
+
+    @property
+    def turn(self) -> str:
+        return self.env.agent_selection
+
+    @property
+    def turn_pretty(self) -> str:
+        return self.turn[:-2]
+
+    @property
+    def observation(self):
+        obs = self.env.last()[0]["observation"]
+        player = obs[:, :, 0].astype("int")
+        opponent = obs[:, :, 1].astype("int")
+        return player - opponent
```

Comment:
So we always see the board relative to the current player.

Not sure this makes the most sense. Would have to think about pro's & con's of this in some depth to properly understand whether this should be changed or not. Let's keep as is for now, but could you make a `TODO` comment mentioning this? Possibly make a ticket too?